File: C_plus_plus/Decode_Ways.cpp
class Solution {
public:
    int record[10000];
    int numDecodings(string s) {
        memset(record,-1,10000*sizeof(int));
        if(s.size()==0) return 0;
        return dp(s,0);
    }
    int dp(string s , int k){
        int a = 0,b = 0;
        if(k == s.size()-1) {
            if(s[k] == '0') return 0;
            return 1;
        }
        if(k == s.size()-2){
            if((s[k]=='1')||((s[k]=='2')&&(s[k+1]<='6'))) {
                if(s[k+1]=='0') return 1;
                return 2;
            }
            if(s[k]!='0'&&s[k+1]!='0') return 1;
            return 0;
        }
        if(s[k]!='0'){
            if(record[k+1]==-1) record[k+1] = dp(s,k+1);
            a = record[k+1];
        }
        if((s[k]=='1')||((s[k]=='2')&&(s[k+1]<='6'))){
            if(record[k+2]==-1) record[k+2] = dp(s,k+2);
            b = record[k+2];
        }
        return a+b;
    }
};

File: java/Decode_Ways.java
// DP solution
// if s.charAt(i) == '0', dp[i] = 0, 
// if Integer.parseInt(s.substring(i,i+2)) <= 26, dp[i] = dp[i+1] + dp[i+2];
// else dp[i] = dp[i+1];
// Time and space are both O(n).

public class Solution {
    public int numDecodings(String s) {
        int n = s.length();
        int[] dp = new int[n+1];
        if (n == 0) return 0;
        dp[n] = 1;
        dp[n-1] = s.charAt(n-1) == '0' ? 0 : 1;
        for (int i = n-2; i >= 0; --i) {
            if (s.charAt(i) == '0') {
                dp[i] = 0;
                continue;
            }
            int two = Integer.parseInt(s.substring(i,i+2));
            dp[i] = dp[i+1];
            if (two <= 26) dp[i] += dp[i+2];
        }
        return dp[0];
    }
}
